File: internal/workspace/paths.go
package workspace

import (
	"os"
	"path/filepath"

	"moyuan-code/internal/fsutil"
)

const DirName = ".moyuan"

type Paths struct {
	RootDir          string
	MoyuanDir        string
	ProjectYAML      string
	RepositoryYAML   string
	AccessYAML       string
	PermissionsJSON  string
	LoggingJSON      string
	AuthDir          string
	LogsDir          string
	LifecycleDir     string
	EpicsDir         string
	IssuesDir        string
	IssueGraphsDir   string
	SchedulesDir     string
	RunsDir          string
	QualityDir       string
	ReviewsDir       string
	ReleasesDir      string
	DeploymentsDir   string
	ComprehensionDir string
	MemoryDir        string
	AgentsDir        string
	RuntimesDir      string
	SkillsDir        string
	TmpDir           string
	LocksDir         string
}

func ForRoot(rootDir string) Paths {
	rootDir, _ = filepath.Abs(rootDir)
	moyuanDir := filepath.Join(rootDir, DirName)
	return Paths{
		RootDir:          rootDir,
		MoyuanDir:        moyuanDir,
		ProjectYAML:      filepath.Join(moyuanDir, "project.yaml"),
		RepositoryYAML:   filepath.Join(moyuanDir, "repository.yaml"),
		AccessYAML:       filepath.Join(moyuanDir, "policies", "access.yaml"),
		PermissionsJSON:  filepath.Join(moyuanDir, "policies", "permissions.json"),
		LoggingJSON:      filepath.Join(moyuanDir, "policies", "logging.json"),
		AuthDir:          filepath.Join(moyuanDir, "auth"),
		LogsDir:          filepath.Join(moyuanDir, "logs"),
		LifecycleDir:     filepath.Join(moyuanDir, "lifecycle"),
		EpicsDir:         filepath.Join(moyuanDir, "lifecycle", "epics"),
		IssuesDir:        filepath.Join(moyuanDir, "lifecycle", "issues"),
		IssueGraphsDir:   filepath.Join(moyuanDir, "lifecycle", "issue-graphs"),
		SchedulesDir:     filepath.Join(moyuanDir, "lifecycle", "schedules"),
		RunsDir:          filepath.Join(moyuanDir, "lifecycle", "runs"),
		QualityDir:       filepath.Join(moyuanDir, "lifecycle", "quality"),
		ReviewsDir:       filepath.Join(moyuanDir, "lifecycle", "reviews"),
		ReleasesDir:      filepath.Join(moyuanDir, "lifecycle", "releases"),
		DeploymentsDir:   filepath.Join(moyuanDir, "lifecycle", "deployments"),
		ComprehensionDir: filepath.Join(moyuanDir, "comprehension"),
		MemoryDir:        filepath.Join(moyuanDir, "memory"),
		AgentsDir:        filepath.Join(moyuanDir, "agents"),
		RuntimesDir:      filepath.Join(moyuanDir, "runtimes"),
		SkillsDir:        filepath.Join(moyuanDir, "skills"),
		TmpDir:           filepath.Join(moyuanDir, "tmp"),
		LocksDir:         filepath.Join(moyuanDir, ".locks"),
	}
}

func ResolveRoot(startDir string) (string, bool) {
	current, _ := filepath.Abs(startDir)
	for {
		if fsutil.Exists(filepath.Join(current, DirName)) {
			return current, true
		}
		parent := filepath.Dir(current)
		if parent == current {
			return "", false
		}
		current = parent
	}
}

func EnsureDirs(paths Paths) error {
	dirs := []string{
		paths.MoyuanDir,
		paths.AuthDir,
		paths.AgentsDir,
		paths.ComprehensionDir,
		paths.EpicsDir,
		paths.IssuesDir,
		paths.IssueGraphsDir,
		paths.SchedulesDir,
		paths.RunsDir,
		paths.QualityDir,
		filepath.Join(paths.QualityDir, "reports"),
		paths.ReviewsDir,
		paths.ReleasesDir,
		paths.DeploymentsDir,
		paths.LogsDir,
		paths.MemoryDir,
		filepath.Join(paths.MoyuanDir, "models"),
		filepath.Join(paths.MoyuanDir, "policies"),
		paths.RuntimesDir,
		paths.SkillsDir,
		paths.TmpDir,
		paths.LocksDir,
	}
	for _, dir := range dirs {
		if err := os.MkdirAll(dir, 0o755); err != nil {
			return err
		}
	}
	return nil
}

File: src/core/paths.js
import fs from 'node:fs/promises';
import path from 'node:path';
import { fileExists } from '../lib/fs.js';

export const MOYUAN_DIR_NAME = '.moyuan';

export function getMoyuanDir(rootDir) {
  return path.join(rootDir, MOYUAN_DIR_NAME);
}

export function getWorkspacePaths(rootDir) {
  const moyuanDir = getMoyuanDir(rootDir);
  return {
    rootDir,
    moyuanDir,
    projectYaml: path.join(moyuanDir, 'project.yaml'),
    repositoryYaml: path.join(moyuanDir, 'repository.yaml'),
    accessYaml: path.join(moyuanDir, 'policies', 'access.yaml'),
    permissionsYaml: path.join(moyuanDir, 'policies', 'permissions.yaml'),
    loggingYaml: path.join(moyuanDir, 'policies', 'logging.yaml'),
    projectDir: path.join(moyuanDir),
    authDir: path.join(moyuanDir, 'auth'),
    logsDir: path.join(moyuanDir, 'logs'),
    lifecycleDir: path.join(moyuanDir, 'lifecycle'),
    epicsDir: path.join(moyuanDir, 'lifecycle', 'epics'),
    issuesDir: path.join(moyuanDir, 'lifecycle', 'issues'),
    issueGraphsDir: path.join(moyuanDir, 'lifecycle', 'issue-graphs'),
    schedulesDir: path.join(moyuanDir, 'lifecycle', 'schedules'),
    runsDir: path.join(moyuanDir, 'lifecycle', 'runs'),
    qualityDir: path.join(moyuanDir, 'lifecycle', 'quality'),
    reviewsDir: path.join(moyuanDir, 'lifecycle', 'reviews'),
    releasesDir: path.join(moyuanDir, 'lifecycle', 'releases'),
    deploymentsDir: path.join(moyuanDir, 'lifecycle', 'deployments'),
    comprehensionDir: path.join(moyuanDir, 'comprehension'),
    memoryDir: path.join(moyuanDir, 'memory'),
    agentsDir: path.join(moyuanDir, 'agents'),
    runtimesDir: path.join(moyuanDir, 'runtimes'),
    skillsDir: path.join(moyuanDir, 'skills'),
    tmpDir: path.join(moyuanDir, 'tmp'),
    locksDir: path.join(moyuanDir, '.locks')
  };
}

export async function resolveWorkspaceRoot(startDir = process.cwd()) {
  let current = path.resolve(startDir);

  while (true) {
    if (await fileExists(getMoyuanDir(current))) {
      return current;
    }

    const parent = path.dirname(current);
    if (parent === current) {
      return null;
    }
    current = parent;
  }
}

export async function ensureWorkspaceDirs(rootDir, dirs) {
  for (const dir of dirs) {
    await fs.mkdir(dir, { recursive: true });
  }
}
